cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/cache/watcher_cache.go

```diff
@@ -0,0 +1,873 @@
+		}
+		close(requestChan)
+	}()
+
+	wg.Wait()
+}
+
+// GetForOne() is like fetchForOne() but if there is a cache miss, it will also check the
+// k8s for the corresponding object, process it and then add it to the cache and return the
+// result.
+func (c *NamespacedResourceWatcherCache) GetForOne(key string) (interface{}, error) {
+	c.resyncCond.L.(*sync.RWMutex).RLock()
+	defer c.resyncCond.L.(*sync.RWMutex).RUnlock()
+
+	log.Infof("+GetForOne(%s)", key)
+	var value interface{}
+	var err error
+	if value, err = c.fetchForOne(key); err != nil {
+		return nil, err
+	} else if value == nil {
+		// cache miss
+		c.queue.Add(key)
+		// now need to wait until this item has been processed by runWorker().
+		// a little bit in-efficient: syncHandler() will eventually call config.onAdd()
+		// which encode the data as []byte before storing it in the cache. That part is fine.
+		// But to get back the original data we have to decode it via config.onGet().
+		// It'd nice if there was a shortcut and skip the cycles spent decoding data from
+		// []byte to repoCacheEntry
+		c.queue.WaitUntilDoneWith(key)
+		// yes, there is a small time window here between after we are done with WaitUntilDoneWith
+		// and the following fetch, where another concurrent goroutine may force the newly added
+		// cache entry out, but that is an edge case and I am willing to overlook it for now
+		// To fix it, would somehow require WaitUntilDoneWith returning a value from a cache, so
+		// the whole thing would be atomic. Don't know how to do this yet
+		return c.fetchForOne(key)
+	}
+	return value, nil
+}
+
+func (c *NamespacedResourceWatcherCache) memoryStats() (used, total string) {
+	used, total = "?", "?"
+	// ref: https://redis.io/commands/info
+	if meminfo, err := c.redisCli.Info(c.redisCli.Context(), "memory").Result(); err == nil {
+		for _, l := range strings.Split(meminfo, "\r\n") {
+			if used == "?" && strings.HasPrefix(l, "used_memory_rss_human:") {
+				used = strings.Split(l, ":")[1]
+			} else if total == "?" && strings.HasPrefix(l, "maxmemory_human:") {
+				total = strings.Split(l, ":")[1]
+			}
+			if used != "?" && total != "?" {
+				break
+			}
+		}
+	} else {
+		log.Warningf("Failed to get redis memory stats due to: %v", err)
+	}
+	return used, total
+}
+
+// this func is used by unit tests only
+func (c *NamespacedResourceWatcherCache) ExpectAdd(key string) {
+	c.queue.ExpectAdd(key)
+}
+
+// this func is used by unit tests only
+func (c *NamespacedResourceWatcherCache) WaitUntilDoneWith(key string) {
+	c.queue.WaitUntilDoneWith(key)
+}
+
+// this func is used by unit tests only
+func (c *NamespacedResourceWatcherCache) ExpectResync() {
+	c.resyncCond.L.Lock()
+	defer c.resyncCond.L.Unlock()
+
+	c.expectResync = true
+}
+
+// this func is used by unit tests only
+func (c *NamespacedResourceWatcherCache) WaitUntilResyncComplete() {
+	c.resyncCond.L.Lock()
+	defer c.resyncCond.L.Unlock()
+
+	for c.expectResync {
+		c.resyncCond.Wait()
+	}
+}
```

Comment:
Geez I feel like going through an interview process where the interviewer asks "what do you do when a team member disagrees with you or questions your choices". Yes, you mentioned this already Michael. I don't know whether this is a criticism or what. I don't quite know how to respond to make you happy. I too, wish things were simple. Having said that, this particular for loop follows exactly the pattern the documentation tells you to do:
https://pkg.go.dev/sync#Cond.Wait
In general, I don't set out to write complicated code. If in the end it happens to be complicated, it's just how it evolved in order to solve some problem.  When have a problem in front of me, I try to solve it the best way I can. Any constructive ideas, such as how to solve a problem in a simpler way, I am generally open to.  Thanks

Reply:
Sorry Greg. I don't mean to leave you feeling that way, nor did I mean to mention it more than once. Sometimes I write comments inline then summarize when submitting the review. The comment here is the inline one when I found at the end of the file that I was really struggling to stay focused any more, but I kept going to the end, and added the review summary. I'll update to stop adding a summary so it doesn't get repeated.

Nor am I trying to disagree with your choices here... as above, it's excellent that you test the code so well as that makes it much safer for anyone to work on, which is great. I am *very* happy that we've got someone who dives in to an issue like this with the goal of finding the best way to solve it, it's excellent. Sometimes a problem is complex and the solution will reflect that, of course.

In terms of reviewing large complex diffs, yep, it's sometimes inevitable, and when they are, it's hard to do a good job as a reviewer (I feel it's hard for me at least). One constructive idea that I find helpful for myself is that of "prequel" (rather than sequel) branches: if I'm working on a PR to do X, but while working on X I find I also need to do Y to be able to do X, where possible (and it's not always), I save my work, create a new branch from main, do Y, push that for review, rebase my X branch on that one and continue while I wait for Y to be reviewed and landed. Once Y is landed I can rebase on main and continue. Not saying that's possible here, nor that it's something you don't do or have not done in the past, just something I find helpful. (Here's an [example prequel PR](https://github.com/kubeapps/kubeapps/pull/3781)).

The day was also a long day of reviews (and only reviews) for me, which was pretty hard to get through. I'll try to make sure next time that I don't just review all day, but instead just set part of the day aside and do them over a couple of days when there are so many.

Reply:
I understand. Thanks for all your comments. Cheers